esp-hal/src/i2c/master/mod.rs

```diff
@@ -1331,47 +1326,72 @@ impl Driver<'_> {
             // Use Most Significant Bit first for sending and receiving data
             w.tx_lsb_first().clear_bit();
             w.rx_lsb_first().clear_bit();
+
             #[cfg(not(esp32))]
             w.arbitration_en().clear_bit();
+
+            #[cfg(esp32s2)]
+            w.ref_always_on().set_bit();
+
             // Ensure that clock is enabled
             w.clk_en().set_bit()
         });
+    }
 
-        #[cfg(esp32s2)]
-        self.regs().ctr().modify(|_, w| w.ref_always_on().set_bit());
+    /// Configures the I2C peripheral with the specified frequency, clocks, and
+    /// optional timeout.
+    fn setup(&self, config: &Config) -> Result<(), ConfigError> {
+        self.init_master();
 
         // Configure filter
         // FIXME if we ever change this we need to adapt `set_frequency` for ESP32
         set_filter(self.regs(), Some(7), Some(7));
 
         // Configure frequency
-        self.set_frequency(config, config.timeout)?;
+        self.set_frequency(config)?;
 
         // Configure additional timeouts
         #[cfg(not(any(esp32, esp32s2)))]
-        self.regs()
-            .scl_st_time_out()
-            .write(|w| unsafe { w.scl_st_to().bits(config.scl_st_timeout.value()) });
-        #[cfg(not(any(esp32, esp32s2)))]
-        self.regs()
-            .scl_main_st_time_out()
-            .write(|w| unsafe { w.scl_main_st_to().bits(config.scl_main_st_timeout.value()) });
-
-        self.update_config();
+        {
+            self.regs()
+                .scl_st_time_out()
+                .write(|w| unsafe { w.scl_st_to().bits(config.scl_st_timeout.value()) });
+            self.regs()
+                .scl_main_st_time_out()
+                .write(|w| unsafe { w.scl_main_st_to().bits(config.scl_main_st_timeout.value()) });
+        }
 
-        // Reset entire peripheral (also resets fifo)
-        self.reset();
+        self.update_registers();
 
         Ok(())
     }
 
     /// Resets the I2C controller (FIFO + FSM + command list)
-    fn reset(&self) {
-        // Reset the FSM
-        // (the option to reset the FSM is not available for the ESP32)
-        #[cfg(not(esp32))]
-        self.regs().ctr().modify(|_, w| w.fsm_rst().set_bit());
-        self.reset_before_transmission();
+    // This function implements esp-idf's `s_i2c_hw_fsm_reset`, without the
+    // clear_bus=true parts.
+    // https://github.com/espressif/esp-idf/blob/27d68f57e6bdd3842cd263585c2c352698a9eda2/components/esp_driver_i2c/i2c_master.c#L115
+    //
+    // Make sure you don't call this function in the middle of a transaction. If the
+    // first command in the command list is not a START, the hardware will hang
+    // with no timeouts.
+    fn reset_fsm(&self) {
+        cfg_if::cfg_if! {
+            if #[cfg(any(esp32c6, esp32h2))] {
```

Comment:
Not for this PR, but maybe we want to start leveraging the cfg symbols we can make into "quirk" flags. I think this will become more important as we add more chips, as adding one may require touching all these cfgs that our outlined like this.

Reply:
Ah yes, I forgot about your versioned peripheral idea! That might also come in handy soon. We can try and eval both approaches when we're looking at adding the C5.